lib/Memoizer.js

```diff
@@ -1,61 +1,117 @@
-  this._memo_all = {};
+
+  if (!ss.storage.queryCache) {
+    ss.storage.queryCache = {};
+  }
   this._onPrefChange = this._onPrefChange.bind(this);
   simplePrefs.on("", this._onPrefChange);
 }
 
 Memoizer.prototype = {
 
+  /**
+   * Return a function that will wrap a given function and cache its results.
+   *
+   * @param {String} key
+   *   a key to store/retrieve cached values
+   *
+   * @param {Function} func
+   *   a function to wrap for caching
+   *
+   * @return {Promise}
+   *   A promise whose result contains data
+   */
   memoize(key, func) {
-    let concat = String.prototype.concat;
     return (...args) => {
-      if (!this._memo_all[key]) {
-        this._memo_all[key] = {};
-      }
-      let memo = this._memo_all[key];
-
       // allow object params to be passed, as is expected of query functions
-      let parts = [];
-      for (let arg of args) {
-        if (typeof arg === "object") {
-          // sort so we have a stable key
-          let props = Object.keys(arg).sort();
-          for (let prop of props) {
-            parts.push(`${prop}:${arg[prop]}`);
-          }
-        } else {
-          parts.push(arg);
-        }
-      }
-      let repr = concat.call(parts);
+      let repr = this._makeRepr(args);
 
-      if (this._cacheEnabled && repr in memo) {
+      if (this._cacheEnabled && this.hasRepr(key, repr)) {
         Services.obs.notifyObservers(null, `${key}-cache`, "hit");
-        return memo[repr];
+        return new Promise(resolve => resolve(this.getData(key, repr)));
       } else {
         Services.obs.notifyObservers(null, `${key}-cache`, "miss");
-        memo[repr] = func.apply(this, args);
-        return memo[repr];
+        return Task.spawn(function*() {
+          let data = yield func.apply(this, args);
+          this.cacheData(key, repr, data);
+          return data;
+        }.bind(this));
       }
     };
   },
 
+  /**
+   * Returns a string representation of parameters to be used as a cache sub-key
+   */
+  _makeRepr(args) {
+    let parts = [];
+    for (let arg of args) {
+      if (typeof arg === "object") {
+        // sort so we have a stable key
+        let props = Object.keys(arg).sort();
+        for (let prop of props) {
+          parts.push(`${prop}:${arg[prop]}`);
+        }
+      } else {
+        parts.push(arg);
+      }
+    }
+    return concat.call(parts);
+  },
+
+  /**
+   * Stores data in the cache
+   *
+   * @param {String} key
+   *   A key for values to be stored
+   * @param {String} repr
+   *   A sub-key for cached values
+   * @param {Object} data
+   *   JSON serializable data to cache
+   */
+  cacheData(key, repr, data) {
+    if (!ss.storage.queryCache[key]) {
+      ss.storage.queryCache[key] = {};
+    }
+    ss.storage.queryCache[key][repr] = data;
+  },
+
+  getData(key, repr) {
+    if (this.hasRepr(key, repr)) {
+      return ss.storage.queryCache[key][repr];
+    }
+    return undefined;
```

Comment:
This line isn't needed since that's the default behavior of functions in js if nothing is returned.